### ISSUE TITLE
OpenTK.Matrix4 Row0,Row1,Row2,Row3 fields are actually the columns

### ISSUE THREAD

**Issue by Electron4** [opentk/opentk#512]
This is how a translation matrix should look like :
![2017-04-09 13_21_31-translation geometry - wikipedia](https://cloud.githubusercontent.com/assets/26382943/24836546/78c737a6-1d27-11e7-8c37-58993e6c6ebb.png)

so with the first 3 elements (x,y,z) of the LAST COLUMN doing the actual translation.

But OpenTK defines it like this:
![2017-04-09 13_16_07-ilspy](https://cloud.githubusercontent.com/assets/26382943/24836542/67f406c0-1d27-11e7-8cb6-5db58ef261e2.png)

by setting the Row3 with the X,Y,Z values, when it actually sets Column3.

So basically the naming of the Row fields is wrong - it should have been "Column" instead of "Row".

Another way to look at this -  OpenGL expects matrices to be given in column major format, so why would the Matrix4 fields be 'rows' ?

This is probably too late to change now, so it may force some devs to write their own Vector and Matrix strcutures, which is a bit sad.

**Comment by winterhell** [comment]
The OpenTK math works fine with the GL Bindings it provides. I'm using GL.UniformMatrix4(location, transpose= false, ref myMatrix4);
If you find that there are some extra transposings done that cause performance drop, you are welcome to post about it.

**Comment by varon** [comment]
@Electron4: The translation matrix you gave is in[ row-major order](https://en.wikipedia.org/wiki/Row-_and_column-major_order). OpenGL uses column-major order, which looks like this:
```
1, 0, 0, 0
0, 1, 0, 0
0, 0, 1, 0
x, y, z, 1
```

They're equivalent representations - they just use different conventions.  OpenGL and by extension, OpenTK, standardizes on column major, rather than row-major of DirectX. You _can_ do your math in either way you like, but I'd generally advise just standardizing on the column-major ordering for your maths, as it's way simpler to just work using that system. Most of our users don't even notice this, because they just use the library functions to create matrices rather than doing it by hand.

Mat*Vec for one is equivalent to Vec*Mat for the other.

See also:
https://stackoverflow.com/questions/13293469/why-does-my-translation-matrix-needs-to-be-transposed

**Comment by Electron4** [comment]
makes sense now, thank you :)

**Comment by mizuno-gsinet** [comment]
@varon Row- or column-major order is about how matrix elements are laid out on the memory and not a property of a matrix.

Column vectors are used thoroughly to represent coordinates in OpenGL spec (and also in OpenGL Programming Guide). So I believe @Electron4's notation should be preferred for representing a translation matrix and the original argument is still valid.

**Comment by varon** [comment]
@mizuno-gsinet  It's not about memory layout; it's a convention relating to what each matrix element means. We could just as well order our vectors WXYZ and adjust the matrix math accordingly to produce the same end result.

Long story short is that the conventions we've established determine the matrix-vector multiplication order the library. Remember that both representations are equivalent and both are correct.

OpenTK does not prescribe matrices take any specific form, so if for some reason you want to do all of your math in a different convention, nothing will prevent you from doing so.

We strongly advise that the library conventions are followed, as they're established, work well, and will save you a ton of headaches if followed.

**Comment by mizuno-gsinet** [comment]
@varon 

>@mizuno-gsinet It's not about memory layout;

I'm sorry but I don't understand what you mean. Let me explain my understanding.

A matrix is merely a table of coefficients of a linear transformation. For example, transformation

```
y1 = a11 * x1 + a12 * x2 + a13 * x3
y2 = a21 * x1 + a22 * x2 + a23 * x3
y3 = a31 * x1 + a32 * x2 + a33 * x3
```

can be written as

```
(y1    (a11 a12 a13  (x1
 y2  =  a21 a22 a23   x2     (y = A x)
 y3)    a31 a32 a33)  x3)
```

or

```
                        (a11 a21 a31
(y1 y2 y3) = (x1 x2 x3)  a12 a22 a32     (y^T = x^T A^T)
                         a13 a23 a33)
```

Different matrix (A or A^T) is used depending on the notation of vectors; column vectors or row vectors. OpenGL documents use the former one.

And there are two ways to lay out elements of a matrix on memory; row-major order and column-major order. Row-major order means to lay out elements of matrix A as follows.

```
a11 a12 a13 a21 a22 a23 a31 a32 a33
```

It also means that elements of matrix A^T are laid out as follows.

```
a11 a21 a31 a12 a22 a32 a13 a23 a33
```

On the other hand, column-major order, which is used in OpenGL, means to lay out elements of matrix A as follows.

```
a11 a21 a31 a12 a22 a32 a13 a23 a33
``` 

Or to lay out elements of matrix A^T as follows.

```
a11 a12 a13 a21 a22 a23 a31 a32 a33
```

Could you point out my mistakes?

>We strongly advise that the library conventions are followed, as they're established, work well, and will save you a ton of headaches if followed.

Thank you for the advice. Actually, I'm already doing so. But I still believe that Row0 should be Column0...

**Comment by mizuno-gsinet** [comment]
Now I think I understand what @varon was telling. So I asked a follow-up question in #922. Thank you.
